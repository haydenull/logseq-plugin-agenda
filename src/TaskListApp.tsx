import React, { useEffect, useState } from 'react'
import { useAtom } from 'jotai'
import { projectSchedulesAtom, todaySubscriptionSchedulesAtom } from '@/model/schedule'
import { categorizeSubscriptions, categorizeTasks } from '@/util/schedule'
import SidebarTask from './components/SidebarTask'
import { fullEventsAtom, journalEventsAtom, projectEventsAtom, todayTasksAtom } from '@/model/events'
import { getInternalEvents } from './util/events'
import SidebarSubscription from './components/SidebarSubscription'
import { renderModalApp } from './main'

const App: React.FC<{
  containerId: string
}> = ({ containerId }) => {

  // TODO: 使用 only-write 减少重新渲染
  const [, setProjectSchedules] = useAtom(projectSchedulesAtom)

  const [todaySubscriptions] = useAtom(todaySubscriptionSchedulesAtom)
  const { allDaySubscriptions, timeSubscriptions } = categorizeSubscriptions(todaySubscriptions)
  const [todayTasks] = useAtom(todayTasksAtom)
  const { overdueTasks, allDayTasks, timeTasks } = categorizeTasks(todayTasks)

  const [, setFullEvents] = useAtom(fullEventsAtom)
  const [, setJournalEvents] = useAtom(journalEventsAtom)
  const [, setProjectEvents] = useAtom(projectEventsAtom)

  useEffect(() => {
    async function fetchSchedules() {
      const res = await getInternalEvents()
      if (res) {
        const { fullEvents, journalEvents, projectEventsMap } = res
        setFullEvents(fullEvents)
        setJournalEvents(journalEvents)
        setProjectEvents(projectEventsMap)
      }
    }
    fetchSchedules()
    logseq.DB.onChanged(({ blocks, txData, txMeta }) => {
      if (txData?.some(item => item?.[1] === 'marker') && parent.document.querySelector('#' + containerId)) {
        fetchSchedules()
      }
    })
  }, [])

  return (
    <main>
      <button onClick={() => {
        renderModalApp({ type: 'addDailyLog' })
        logseq.showMainUI()
      }}>Add Daily Log</button>
      {
        overdueTasks?.length === 0 && allDayTasks?.length === 0 && timeTasks?.length === 0 && (
          <div>Agenda: No Task Today</div>
        )
      }
      {
        overdueTasks.length > 0 && (
          <div style={{ margin: '8px 0' }}>
            {/* <span>Overdue</span> */}
            {
              overdueTasks.map(task => (
                <SidebarTask key={task.id} task={task} type="overdue" />
              ))
            }
          </div>
        )
      }
      {
        (allDayTasks.length > 0 || allDaySubscriptions.length > 0) && (
          <div>
            {/* <span>All Day</span> */}
            {
              allDaySubscriptions.map(subscription => (
                <SidebarSubscription key={subscription.id} subscription={subscription} type="allDay" />
              ))
            }
            {
              allDayTasks.map(task => (
                <SidebarTask key={task.id} task={task} type="allDay" />
              ))
            }
          </div>
        )
      }
      {
        (timeTasks.length > 0 || timeSubscriptions.length > 0) && (
          <div>
            {/* <span>Time</span> */}
            {
              timeSubscriptions.map(subscription => (
                <SidebarSubscription key={subscription.id} subscription={subscription} type="time" />
              ))
            }
            {
              timeTasks.map(task => (
                <SidebarTask key={task.id} task={task} type="time" />
              ))
            }
          </div>
        )
      }
    </main>
  )
}

export default App
