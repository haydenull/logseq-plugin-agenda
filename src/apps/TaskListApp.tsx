import clsx from 'clsx'
import dayjs, { type Dayjs } from 'dayjs'
import { useAtom } from 'jotai'
import React, { useEffect, useState } from 'react'
import { IoIosArrowBack, IoIosArrowForward } from 'react-icons/io'
import { MdAddTask } from 'react-icons/md'
import { TbActivity } from 'react-icons/tb'

import SidebarSubscription from '@/components/SidebarSubscription'
import SidebarTask from '@/components/SidebarTask'
import { renderModalApp } from '@/main'
import { fullEventsAtom, journalEventsAtom, projectEventsAtom, todayTasksAtom } from '@/model/events'
import { todaySubscriptionSchedulesAtom, subscriptionSchedulesAtom } from '@/model/schedule'
import { getInitialSettings } from '@/util/baseInfo'
import { getInternalEvents, getTasksInTimeRange } from '@/util/events'
import { categorizeSubscriptions, categorizeTasks } from '@/util/schedule'
import { getSubscriptionsInTimeRange } from '@/util/subscription'
import { genWeekDays } from '@/util/util'

const App: React.FC<{
  containerId: string
}> = ({ containerId }) => {
  const [, setJournalEvents] = useAtom(journalEventsAtom)
  const [, setProjectEvents] = useAtom(projectEventsAtom)

  // calendar
  const [weekDays, setWeekDays] = useState<Dayjs[]>([])
  const [activeDay, setActiveDay] = useState<Dayjs>(dayjs())

  // subscriptions
  const [fullSubscriptions] = useAtom(subscriptionSchedulesAtom)
  const weekSubscriptionsMap = getSubscriptionsInTimeRange(fullSubscriptions, weekDays)
  const activeDaySubscriptions = weekSubscriptionsMap.get(activeDay.format('YYYY-MM-DD')) ?? []
  const { allDaySubscriptions, timeSubscriptions } = categorizeSubscriptions(activeDaySubscriptions)

  // tasks
  const [fullTasks, setFullEvents] = useAtom(fullEventsAtom)
  const weekTasksMap = getTasksInTimeRange(fullTasks?.tasks?.withTime, weekDays)
  const activeDayTasks = weekTasksMap.get(activeDay.format('YYYY-MM-DD')) ?? []
  const { overdueTasks, allDayTasks, timeTasks } = categorizeTasks(activeDayTasks)

  // weekday with dots
  const weekDaysWithDots = Object.fromEntries(
    weekDays.map((day) => {
      const dayString = day.format('YYYY-MM-DD')
      const subscriptions = weekSubscriptionsMap.get(dayString) ?? []
      const tasks = weekTasksMap.get(dayString) ?? []
      return [dayString, subscriptions.length > 0 || tasks.length > 0]
    }),
  )

  const onClickCalendarNumber = (day: Dayjs) => {
    setActiveDay(day)
  }
  const onClickCalendarArrow = (direction: 'back' | 'forward') => {
    const oneDay = direction === 'back' ? weekDays[0]?.add(-1, 'day') : weekDays[weekDays.length - 1].add(1, 'day')
    const { weekStartDay } = getInitialSettings()
    const days = genWeekDays(weekStartDay, oneDay)
    setWeekDays(days)
    setActiveDay((_old) => days.find((day) => day.day() === _old.day()) ?? days[0])
  }

  useEffect(() => {
    async function fetchSchedules() {
      const res = await getInternalEvents()
      if (res) {
        const { fullEvents, journalEvents, projectEventsMap } = res
        setFullEvents(fullEvents)
        setJournalEvents(journalEvents)
        setProjectEvents(projectEventsMap)
      }
    }
    fetchSchedules()
    logseq.DB.onChanged(({ blocks, txData, txMeta }) => {
      if (txData?.some((item) => item?.[1] === 'marker') && parent.document.querySelector('#' + containerId)) {
        fetchSchedules()
      }
    })
  }, [])

  useEffect(() => {
    const { weekStartDay } = getInitialSettings()
    const days = genWeekDays(weekStartDay)
    setWeekDays(days)
  }, [])

  return (
    <main>
      {/* ========= toolbar start ========= */}
      <div className="flex items-center">
        <a
          className="button items-center"
          style={{ display: 'flex' }}
          title="Add Daily Log"
          href="javascript:void(0);"
          onClick={() => {
            renderModalApp({ type: 'addDailyLog' })
            logseq.showMainUI()
          }}
        >
          <TbActivity />
        </a>
        <a
          className="button items-center"
          style={{ display: 'flex' }}
          title="Create Schedule"
          href="javascript:void(0);"
          onClick={() => {
            renderModalApp({ type: 'modifySchedule', data: { type: 'create' } })
            logseq.showMainUI()
          }}
        >
          <MdAddTask />
        </a>
      </div>
      {/* ========= toolbar end ========= */}

      {/* ========= calendar start ========= */}
      <div className="flex items-center justify-between agenda-sidebar-calendar">
        <div
          className="agenda-sidebar-calendar__number flex items-center justify-center"
          onClick={() => onClickCalendarArrow('back')}
        >
          <IoIosArrowBack />
        </div>
        {weekDays.map((day) => (
          <div
            className="flex flex-col items-center"
            key={day.format('YYYY-MM-DD')}
            onClick={() => onClickCalendarNumber(day)}
          >
            <span
              className={clsx({
                'agenda-sidebar-calendar__week-day--today': day.isSame(dayjs(), 'day'),
              })}
            >
              {day.format('dd')?.charAt(0)}
            </span>
            <div
              className={clsx('agenda-sidebar-calendar__number flex items-center justify-center', {
                'bg-indigo-600 agenda-sidebar-calendar__number--active': day.isSame(activeDay, 'day'),
                'agenda-sidebar-calendar__number--dot': weekDaysWithDots[day.format('YYYY-MM-DD')],
                'agenda-sidebar-calendar__number--today': day.isSame(dayjs(), 'day'),
              })}
              style={{ opacity: day.isSameOrAfter(dayjs(), 'day') || day.isSame(activeDay, 'day') ? 0.9 : 0.4 }}
            >
              {day.format('DD')}
            </div>
          </div>
        ))}
        <div
          className="agenda-sidebar-calendar__number flex items-center justify-center"
          onClick={() => onClickCalendarArrow('forward')}
        >
          <IoIosArrowForward />
        </div>
      </div>
      {/* ========= calendar end ========= */}

      {overdueTasks?.length === 0 &&
      allDayTasks?.length === 0 &&
      timeTasks?.length === 0 &&
      allDaySubscriptions?.length === 0 &&
      timeSubscriptions?.length === 0 ? (
        <div className="flex justify-center">No tasks found, enjoy your day</div>
      ) : null}
      {/* overdue tasks */}
      {overdueTasks.length > 0 && (
        <div style={{ margin: '8px 0' }}>
          {/* <span>Overdue</span> */}
          {overdueTasks.map((task) => (
            <SidebarTask key={task.id} task={task} type="overdue" />
          ))}
        </div>
      )}
      {/* all day tasks */}
      {(allDayTasks.length > 0 || allDaySubscriptions.length > 0) && (
        <div>
          {/* <span>All Day</span> */}
          {allDaySubscriptions.map((subscription) => (
            <SidebarSubscription key={subscription.id} subscription={subscription} type="allDay" />
          ))}
          {allDayTasks.map((task) => (
            <SidebarTask key={task.id} task={task} type="allDay" />
          ))}
        </div>
      )}
      {/* time tasks */}
      {(timeTasks.length > 0 || timeSubscriptions.length > 0) && (
        <div>
          {/* <span>Time</span> */}
          {timeSubscriptions.map((subscription) => (
            <SidebarSubscription key={subscription.id} subscription={subscription} type="time" />
          ))}
          {timeTasks.map((task) => (
            <SidebarTask key={task.id} task={task} type="time" />
          ))}
        </div>
      )}
    </main>
  )
}

export default App
