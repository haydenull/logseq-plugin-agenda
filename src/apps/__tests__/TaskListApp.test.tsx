import { render, fireEvent, waitForElementToBeRemoved, screen } from '@testing-library/react'
import { beforeEach, describe, expect, test, vi } from 'vitest'

import TaskListApp from '../TaskListApp'

describe('Sidebar App', () => {
  const containerId = 'agenda-task-list-my-test-slot'

  // beforeEach(() => {
  //   vi.mock('logseq', () => {
  //     return {
  //       DB: {
  //         datascriptQuery: () => Promise.resolve([]),
  //         onChanged: (callback) => void 0,
  //         getCurrentPage: () => Promise.resolve({}),
  //       },
  //     }
  //   })
  // })

  test('should render', async () => {
    const sidebarApp = render(<TaskListApp containerId={containerId} />)
    expect(sidebarApp).toMatchSnapshot()
  })

  test('should show overdue tasks in today', async () => {
    render(<TaskListApp containerId={containerId} />)
    // wait data loaded
    // screen.debug()
    await waitForElementToBeRemoved(() => screen.getByText('No tasks found, enjoy your day'), { timeout: 1000 })
    expect(screen.getByText('overdue task')).toBeInTheDocument()
  })
})
