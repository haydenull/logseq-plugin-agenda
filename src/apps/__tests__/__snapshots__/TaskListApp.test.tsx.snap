// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar App > should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main>
        <div
          class="flex items-center"
        >
          <a
            class="button items-center"
            href="javascript:void(0);"
            style="display: flex;"
            title="Add Daily Log"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <desc />
              <path
                d="M0 0h24v24H0z"
                fill="none"
                stroke="none"
              />
              <path
                d="M3 12h4l3 8l4 -16l3 8h4"
              />
            </svg>
          </a>
          <a
            class="button items-center"
            href="javascript:void(0);"
            style="display: flex;"
            title="Create Schedule"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M22 5.18L10.59 16.6l-4.24-4.24 1.41-1.41 2.83 2.83 10-10L22 5.18zM12 20c-4.41 0-8-3.59-8-8s3.59-8 8-8c1.57 0 3.04.46 4.28 1.25l1.45-1.45A10.02 10.02 0 0012 2C6.48 2 2 6.48 2 12s4.48 10 10 10c1.73 0 3.36-.44 4.78-1.22l-1.5-1.5c-1 .46-2.11.72-3.28.72zm7-5h-3v2h3v3h2v-3h3v-2h-3v-3h-2v3z"
              />
            </svg>
          </a>
        </div>
        <div
          class="flex items-center justify-between agenda-sidebar-calendar"
        >
          <div
            class="agenda-sidebar-calendar__number flex items-center justify-center"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
              />
            </svg>
          </div>
          <div
            class="flex flex-col items-center"
          >
            <span
              class="agenda-sidebar-calendar__week-day--today"
            >
              S
            </span>
            <div
              class="agenda-sidebar-calendar__number flex items-center justify-center bg-indigo-600 agenda-sidebar-calendar__number--active agenda-sidebar-calendar__number--today"
              style="opacity: 0.9;"
            >
              27
            </div>
          </div>
          <div
            class="flex flex-col items-center"
          >
            <span
              class=""
            >
              M
            </span>
            <div
              class="agenda-sidebar-calendar__number flex items-center justify-center"
              style="opacity: 0.9;"
            >
              28
            </div>
          </div>
          <div
            class="flex flex-col items-center"
          >
            <span
              class=""
            >
              T
            </span>
            <div
              class="agenda-sidebar-calendar__number flex items-center justify-center"
              style="opacity: 0.9;"
            >
              29
            </div>
          </div>
          <div
            class="flex flex-col items-center"
          >
            <span
              class=""
            >
              W
            </span>
            <div
              class="agenda-sidebar-calendar__number flex items-center justify-center"
              style="opacity: 0.9;"
            >
              30
            </div>
          </div>
          <div
            class="flex flex-col items-center"
          >
            <span
              class=""
            >
              T
            </span>
            <div
              class="agenda-sidebar-calendar__number flex items-center justify-center"
              style="opacity: 0.9;"
            >
              31
            </div>
          </div>
          <div
            class="flex flex-col items-center"
          >
            <span
              class=""
            >
              F
            </span>
            <div
              class="agenda-sidebar-calendar__number flex items-center justify-center"
              style="opacity: 0.9;"
            >
              01
            </div>
          </div>
          <div
            class="flex flex-col items-center"
          >
            <span
              class=""
            >
              S
            </span>
            <div
              class="agenda-sidebar-calendar__number flex items-center justify-center"
              style="opacity: 0.9;"
            >
              02
            </div>
          </div>
          <div
            class="agenda-sidebar-calendar__number flex items-center justify-center"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M294.1 256L167 129c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.3 34 0L345 239c9.1 9.1 9.3 23.7.7 33.1L201.1 417c-4.7 4.7-10.9 7-17 7s-12.3-2.3-17-7c-9.4-9.4-9.4-24.6 0-33.9l127-127.1z"
              />
            </svg>
          </div>
        </div>
        <div
          class="flex justify-center"
        >
          No tasks found, enjoy your day
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main>
      <div
        class="flex items-center"
      >
        <a
          class="button items-center"
          href="javascript:void(0);"
          style="display: flex;"
          title="Add Daily Log"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <desc />
            <path
              d="M0 0h24v24H0z"
              fill="none"
              stroke="none"
            />
            <path
              d="M3 12h4l3 8l4 -16l3 8h4"
            />
          </svg>
        </a>
        <a
          class="button items-center"
          href="javascript:void(0);"
          style="display: flex;"
          title="Create Schedule"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M22 5.18L10.59 16.6l-4.24-4.24 1.41-1.41 2.83 2.83 10-10L22 5.18zM12 20c-4.41 0-8-3.59-8-8s3.59-8 8-8c1.57 0 3.04.46 4.28 1.25l1.45-1.45A10.02 10.02 0 0012 2C6.48 2 2 6.48 2 12s4.48 10 10 10c1.73 0 3.36-.44 4.78-1.22l-1.5-1.5c-1 .46-2.11.72-3.28.72zm7-5h-3v2h3v3h2v-3h3v-2h-3v-3h-2v3z"
            />
          </svg>
        </a>
      </div>
      <div
        class="flex items-center justify-between agenda-sidebar-calendar"
      >
        <div
          class="agenda-sidebar-calendar__number flex items-center justify-center"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
            />
          </svg>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <span
            class="agenda-sidebar-calendar__week-day--today"
          >
            S
          </span>
          <div
            class="agenda-sidebar-calendar__number flex items-center justify-center bg-indigo-600 agenda-sidebar-calendar__number--active agenda-sidebar-calendar__number--today"
            style="opacity: 0.9;"
          >
            27
          </div>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <span
            class=""
          >
            M
          </span>
          <div
            class="agenda-sidebar-calendar__number flex items-center justify-center"
            style="opacity: 0.9;"
          >
            28
          </div>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <span
            class=""
          >
            T
          </span>
          <div
            class="agenda-sidebar-calendar__number flex items-center justify-center"
            style="opacity: 0.9;"
          >
            29
          </div>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <span
            class=""
          >
            W
          </span>
          <div
            class="agenda-sidebar-calendar__number flex items-center justify-center"
            style="opacity: 0.9;"
          >
            30
          </div>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <span
            class=""
          >
            T
          </span>
          <div
            class="agenda-sidebar-calendar__number flex items-center justify-center"
            style="opacity: 0.9;"
          >
            31
          </div>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <span
            class=""
          >
            F
          </span>
          <div
            class="agenda-sidebar-calendar__number flex items-center justify-center"
            style="opacity: 0.9;"
          >
            01
          </div>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <span
            class=""
          >
            S
          </span>
          <div
            class="agenda-sidebar-calendar__number flex items-center justify-center"
            style="opacity: 0.9;"
          >
            02
          </div>
        </div>
        <div
          class="agenda-sidebar-calendar__number flex items-center justify-center"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M294.1 256L167 129c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.3 34 0L345 239c9.1 9.1 9.3 23.7.7 33.1L201.1 417c-4.7 4.7-10.9 7-17 7s-12.3-2.3-17-7c-9.4-9.4-9.4-24.6 0-33.9l127-127.1z"
            />
          </svg>
        </div>
      </div>
      <div
        class="flex justify-center"
      >
        No tasks found, enjoy your day
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
